algorithms/src/traits/snark.rs

```diff
@@ -87,44 +95,42 @@ pub trait SNARK {
 
     fn prove_batch<C: ConstraintSynthesizer<Self::ScalarField>, R: Rng + CryptoRng>(
         fs_parameters: &Self::FSParameters,
-        proving_key: &Self::ProvingKey,
-        input_and_witness: &[C],
+        keys_to_constraints: &BTreeMap<&Self::ProvingKey, &[&C]>,
         rng: &mut R,
     ) -> Result<Self::Proof, SNARKError> {
-        Self::prove_batch_with_terminator(fs_parameters, proving_key, input_and_witness, &AtomicBool::new(false), rng)
+        Self::prove_batch_with_terminator(fs_parameters, keys_to_constraints, &AtomicBool::new(false), rng)
     }
 
     fn prove<C: ConstraintSynthesizer<Self::ScalarField>, R: Rng + CryptoRng>(
         fs_parameters: &Self::FSParameters,
         proving_key: &Self::ProvingKey,
-        input_and_witness: &C,
+        constraints: &C,
         rng: &mut R,
     ) -> Result<Self::Proof, SNARKError> {
-        Self::prove_batch(fs_parameters, proving_key, std::slice::from_ref(input_and_witness), rng)
+        let mut keys_to_constraints = BTreeMap::new();
+        let constraints_vec = vec![constraints];
+        keys_to_constraints.insert(proving_key, constraints_vec.as_slice());
```

Comment:
This doesn't work unfortunately, same for the comments below
```
expected reference `&BTreeMap<&<Self as traits::snark::SNARK>::ProvingKey, &[&_]>`
   found reference `&BTreeMap<&<Self as traits::snark::SNARK>::ProvingKey, &[&C; 1]>
```

Reply:
that's a weird signature; perhaps we could change it so that it ends up being `BTreeMap<&<Self as traits::snark::SNARK>::ProvingKey, &C>` instead?

Reply:
alternatively (though slightly worse, as there is no real need for it other than the current value type), using a 1-element array would be preferable to allocating

Reply:
> that's a weird signature; perhaps we could change it so that it ends up being BTreeMap<&<Self as traits::snark::SNARK>::ProvingKey, &C> instead?

I think that would be too involved at this stage, I wrote it down to look at when working on getting this integrated higher up the stack! That will also give me a good sense of what objects we'll be passing from the compiler.

> alternatively (though slightly worse, as there is no real need for it other than the current value type), using a 1-element array would be preferable to allocating

2bd12df2e